src/djview.1
.\" t
.\"
.\" Copyright (c) 2007- Leon Bottou
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual. Otherwise check the web site
.\" of the Free Software Foundation at http://www.fsf.org.
.TH DJVIEW4 1 "10/11/2001" "DjVuLibre" "DjVuLibre"
.de SS
.SH \\0\\0\\0\\$*
..
.SH NAME
djview4 \- Standalone DjVu viewer

.SH SYNOPSIS
.BI "djview4 [" "options" "] [" "argument" "]"

.SH DESCRIPTION
Standalone viewer for DjVu files.  
Features include navigating documents,
zooming and panning page images, 
producing and displaying thumbnails,
displaying document outlines,
searching documents for particular words in the hidden text layer,
copying hidden text to the clipboard,
saving pages and documents
as bundled or indirect multi-page files, 
and printing page and documents.
The viewer can simultaneously display several pages
using a side-by-side or a continuous layout.

.SH COMMAND LINE ARGUMENT AND OPTIONS
This program
can run as a standalone program or 
as a slave process for the DjVu browser plugin
.BR nsdejavu .
When running as a standalone program,
the command line argument
.I argument
can be:
.IP "*" 3
The filename of a valid DjVu document.
.IP "*" 3
A local DjVu document 
.SM URL
of the form:
.IP "" 5
.BI "file:///" path "/" name.djvu "[?djvuopts&" keyword = value "&...]"
.IP "" 3
The square brackets delimit the optional components of the 
.SM URL.
Various options can be specified using a syntax
similar to that of
.SM CGI
arguments.
Specifying options in this manner is very useful 
for a browser plugin because there are no 
command line arguments.
In the case of a standalone viewer,
all options can be specified as command line arguments.
.IP "*" 3
An remote DjVu document 
.SM URL
of the form:
.IP "" 5
.BI http:// host / path "/" name.djvu "[?djvuopts&" keyword = value "&...]"
.IP "" 5
.BI https:// host / path "/" name.djvu "[?djvuopts&" keyword = value "&...]"
.IP "" 3
Browsing remote DjVu documents with the standalone viewer
is less efficient than using the browser plugin.
The standalone viewer does not benefit from the browser
caching strategies and proxy settings.
Proxy settings for the standalone viewer can be set
independently with the preferences dialog.
.PP

An extensive list of options are recognized.
Most options can be specified as command line argument starting
with the customary dash character
.BR "" "(" "-" ")"
or using the syntax of
.SM CGI
arguments in the document 
.SM URL.
Some options however are only meaningful 
as command line arguments.
Other options are only recognized when
running the 
.SM X11
version of the 
.BR djview4
program.

.SH COMMAND LINE OPTIONS
The following options are only meaningful
when specified on the command line.

.TP
.B "-help"
Display a brief help message.

.TP
.B "-verbose"
Prints informational messages on the console.
This option is very useful because it displays
messages about the unrecognized constructs
in the DjVu annotation and hyperlink layers.

.TP
.BI "-fullscreen, -fs "
Start 
.B djview4 
in full screen mode.
Use the key
.B ESC
to exit the full screen mode.

.TP
.BI "-slideshow[=" arg "]
Start 
.B djview4 
in slide show mode.
The optional numerical argument 
.I arg
specifies the number of seconds
before turning each page.
Use the key
.B ESC
to exit the slide show mode.


.TP
.BI "-style=" "stylename"
Specify the graphical user interface style.
The recognized values for
.I stylename
depend on the installed version of the Qt toolkit.
Common style names include
.BR cde ,
.BR motif ,
.BR plastique ,
.BR platinum ,
and
.BR windows .


.SH X11 OPTIONS
The following command line options are recognized by the
.SM X11
version of the 
.B djview4
program.
Unlike most 
.B djview4 options, 
X11 options that demand an argument
do not use the equal character 
to introduce their argument.
.TP
.BI "-display " "displayname"
Specify that the 
.B djview4
windows should appear on the X11 display 
.IR displayname .

.TP
.BI "-geometry " W x H + X + Y
Specify the initial size and position of the first window
using the traditional X11 geometry specification syntax.
The numerical arguments 
.IR W " and " H
represent the initial window width and height.
The numerical arguments 
.IR X " and " Y
indicate the window position relative to the 
top left corner of the screen.

.TP
.BI "-name " "name"
Set the application name.

.TP
.BI "-title " "title"
Set the title of the first window.

.TP
.BI "-fn " "fontname" ", -font " "fontname"
Specify the name of the default font 
used for buttons and menus.
The font should be specified using 
a X logical font description string.
.TP
.BI "-bg " "color" ", -background " "color"
Specify the default background color for 
graphical user interface elements.
The color should be given as a standard X11 color name.

.TP
.BI "-fg " "color" ", -foreground " "color"
Specify the default foreground color for 
graphical user interface elements.
The color should be given as a standard X11 color name.

.TP
.BI "-btn " "color"
Specify the default button color.
The color should be given as a standard X11 color name.

.TP
.BI "-ncols " "count"
Limit the number of colors allocated on a 8 bit display. 
The default color cube contains 216 distinct colors.
.TP
.B "-cmap"
Force the allocation of a private color map on a 8-bit display.
This might increase the color quality but cause flashing
when the viewer window gets activated.

.SH GENERAL OPTIONS
The following options can be specified as command line options
or can be passed by augmenting the document 
.SM URL
using a syntax similar to that of
.SM CGI
arguments
.IP "" 3
.BI "http://.../file.djvu?djvuopts&" "key=value" "&" "key=value" "&" "..."
.PP
In order to separate real
.SM CGI 
arguments from these options,
the viewer only recognizes keywords 
that appear after the word
.BR djvuopts .
The keywords
.I key
are derived from the option names 
by removing the initial dashes. 

.TP
.BI "-page=" "pagename"
Display a specific document page.
The viewer first searches a page 
whose identifier matches the argument
.IR pagename .
Otherwise, if 
.I pagename
is a number preceded by character 
.BR "+" " or " "-" ","
the viewer performs a displacement relative to the current page.
Otherwise, starting from the current page
and wrapping around, it searches for a page
whose title matches the argument
.IR pagename .
Otherwise, if 
.I pagename
is numerical, it is interpreted as 
an ordinal page number.
Otherwise, and finally, the viewer searches a page
whose name matches
.IR pagename .
.TP
.BI "-pageno=" "pagenumber"
The page searching algorithm for option
.B "page"
can cause ambiguities when page titles can
be interpreted as numbers.
The argument of option
.B "pageno"
is always interpreted as an ordinal page number.
This option is less portable than
.B "page"
because it is not recognized by earlier 
versions of the djvu plugin. 
When using this option is necessary,
it is advisable to use both the
.B "page" 
and 
.B "pageno"
options.
.TP
.BI "-zoom=" zoomfactor
Specify the initial zoom factor.
Unless the toolbar, pop-up menu and keyboard are disabled,
the user will be able to change the zoom factor.
Legal values for
.I zoomfactor
are shown in the below:
.br
.TS
center,box;
lfI l
lfB l
lfB l
lfB l
lfB l
.
number	Magnification factor in range 10% to 999%.
one2one	Select the "one-to-one" mode.
width	Select the "fit width" mode.
page	Select the "fit page" mode.
stretch	Stretch the image to the plugin window size.
.TE
.PP

.TP
.BI "-showposition=" px "," py
Specify a point in the current page that should be
as close as possible to the center of the window.
The horizontal and vertical positions 
.IB  px "," py 
in the current page are given as fractions in range 0 to 1. 
For instance,
.B 0,0
designates the upper left corner of the page,
.B 0.5,0.5
is the center, and 
.B 1,1 
is the lower right corner.

.TP
.BI "-mode=" modespec
Specify the initial display mode.
Unless the toolbar and pop-up menu are disabled,
the user will be able to change it.
Legal values for
.I modespec
are shown in the below:
.br
.TS
center,box;
lfB l
lfB l
lfB l
lfB l
lfB l
.
color	Display the color image.
bw	Display the foreground mask only.
fore	Display the foreground only.
back	Display the background only.
text	Overlay the hidden text over the color image.
.TE
.PP

.TP
.BI "-hor_align=" keyword ", -halign=" keyword
Specify the horizontal position of the page in the viewer window.
(This does not specify what part of the page will be shown, but rather
how margins will be laid out around the page in the plugin window.)
Argument
.I keyword
must be
.BR left ,
.BR center ,
or
.BR right .

.TP
.BI "-ver_align=" keyword ", -valign=" keyword
Specify the vertical position of the page in the viewer window.
(This does not specify what part of the page will be shown, but rather
how margins will be laid out around the page in the plugin window.)
Argument
.I keyword
must be
.BR top ,
.BR center ,
or
.BR bottom .

.TP
.BI "-cache=(yes|no)"
Enable or disable the caching of fully decoded pages of the document.
Caching is on by default.  
Caching of documents whose
.SM URL
does not contain an extension
.BR .djvu " or " .djv
is off by default.

.TP
.BI "-continuous=(yes|no)"
Enable or disable the continuous layout of
multipage documents.

.TP
.BI "-sidebyside=(yes|no), -side_by_side=(yes|no)"
Enable or disable the side-by-side layout of
multipage documents.

.TP
.BI "-coverpage=(yes|no)"
Specify whether the cover page must be displayed alone
when multipage documents are shown in side-by-side layout.

.TP
.BI "-righttoleft=(yes|no)"
Specify whether pages should be arranged right-to-left
when multipage documents are shown in side-by-side layout.

.TP
.BI "-layout=" keyword "{," keyword "}"
Specify the layout settings using a single list
of comma-separated keywords.
The following keywords are recognized:
.br
.TS
center,box;
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
.
single	Disable the side-by-side and continuous modes.
double	Enable the side-by-side mode.
continuous	Enable the continuous mode.
cover,nocover	First page treatment in side-by-side mode.
ltor,rtol	Layout direction for side-by-side mode.
gap,nogap	Specify whether there is a gap between pages.
.TE

.TP
.BI "-scrollbars=(yes|no)"
Enable or disable the presence of scroll bars when the full
image size exceeds the plugin window size. 
The default is yes.

.TP
.BI "-frame=(yes|no)"
Enable or disable the display of a thin frame 
and shadow around the DjVu images.
Frames are enabled by default.

.TP
.BI "-background=" color
Specify the color of the background border
displayed around the document. The color
.I color
must be given in hexadecimal
.SM RRGGBB
or
.SM #RRGGBB
format.

.TP
.BI "-toolbar=" keyword "{(,|+|-)" keyword "}"
Controls the appearance and the contents of the toolbar.
The argument of option
.B toolbar
is composed of a number of keywords separated
by characters comma, plus or minus.
The appearance of the toolbar is controlled by keywords
placed before the first occurrence of a character plus
or minus. The following keywords are recognized in
this context:
.br
.TS
center,box;
lfB l
lfB l
lfB l
lfB l
lfB l
.
no	Disable toolbar.
always	Displays the toolbar.
auto	Enable toolbar "autohide" mode (not implemented).
top	Place toolbar along the top edge.
bottom	Place toolbar along the bottom edge.
.TE
.IP ""
The contents of the toolbar is controlled by keywords
placed after the first occurrence of a
character plus or minus.
Each keyword adds (after a plus) or removes (after a minus)
a particular toolbar button or group of buttons.
The initial content of the toolbar is determined
by the first occurrence of a character plus or minus.
When this is a plus,
the toolbar is initially empty.
When this is a minus,
the toolbar initially contains 
the default selection of buttons.

The following keywords are recognized:
.br
.TS
center,box;
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
.
modecombo	for the display mode selection tool.
zoomcombo	for the zoom selection tool.
zoom	for the zoom buttons.
select	for the selection button.
rotate	for the image rotation buttons.
find	for the text search button.
new	for the new window button.
open	for the open new document button.
save	for the save button.
print	for the print button.
layout	for the page layout buttons.
pagecombo	for the page selection tool.
firstlast	for the first-page and last-page buttons.
prevnext	for the previous- and next-page buttons.
backforw	reserved for the back and forward buttons.
help	for the contextual help button.
.TE
.PP
For the sake of backward compatibility,
the keywords
.BR fore , 
.BR fore_button ,
.BR back ,
.BR back_button ,
.BR bw ,
.BR bw_button ,
.BR color ,
and
.BR color_button
are interpreted like keyword
.BR modecombo ;
the keyword
.BR rescombo
is a synonym of 
.BR zoomcombo ;
the keywords
.BR pan , 
.BR zoomsel ,
and
.BR textsel 
are interpreted like keyword
.BR select ;
and the keyword 
.BR doublepage
is interpreted like keyword
.BR layout .
All other keywords are ignored.
.PP

.TP
.BI "-menubar=(yes|no)"
Enable or disable the presence of the menu bar
located on top of the window.

.TP
.BI "-statusbar=(yes|no)"
Enable or disable the presence of the status bar
located at the bottom of the window.

.TP
.BI "-sidebar=" keyword "{," keyword "}"
Control the dockable panels.
The argument is a comma separated list of keywords.
A first group of keywords selects which panels are affected.
Omitting these keywords selects all panels.
A second group of keywords then controls the visibility and 
the position of the selected panels.
.br
.TS
center,box;
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
.
thumbnails	specify the thumbnail panel.
outline,bookmarks	specify the outline panel.
search,find	specify the search panel.
=
yes,true	show the specified panels (default).
no,false	hide the specified panels.
left	dock specified panels on the left side.
right	dock specified panels on the right side.
top	dock specified panels on the top side.
bottom	dock specified panels on the bottom side.
.TE
.PP
.TP
.BI "-thumbnails=" keyword "{," keyword "}"
Compatibility alias for 
.BI "-sidebar=" keyword "{," keyword "},thumbnails."
.TP
.BI "-outline=" keyword "{," keyword "}"
Compatibility alias for 
.BI "-sidebar=" keyword "{," keyword "},outline."
.TP
.BI "-menu=(yes|no)"
Enable or disable the pop-up menu.

.TP
.BI "-keyboard=(yes|no)"
Enable or disable the DjVu plugin keyboard shortcuts.
The default is yes (enabled). 
Note that disabling the keyboard handling 
does not disable the shortcuts associated
with active menus (see options 
.BR "-menu" " and " "-menubar" "),"
visible toolbars and sidebars (see options
.BR "-toolbar" " and " "-sidebar" ")." 

.TP
.BI "-mouse=(yes|no)"
Enable or disable mouse interaction for panning and selecting.
The default is yes (enabled).

.TP
.BI "-links=(yes|no)"
Enable or disable hyper-links in the DjVu image.
Hyper-links are enabled by default.

.TP
.BI "-highlight=" x "," y "," w "," h "[," color "]"
Display a highlighted rectangle at the specified coordinates in
the current page and with the specified color.  Coordinates
.IR x ", " y ", " w ", and " h
are measured in document image coordinates (not screen coordinates).
The origin is set at the bottom left corner of the image.  The color
.I color
must be given in hexadecimal
.SM RRGGBB
or
.SM #RRGGBB
format.
Multiple highlighted zone can be specified 
and can be interspersed with multiple
.BI "-page=" "pagename"
options. 
.TP
.BI "-find=" text
Highlight occurrences of the given string 
.IR text .
This option works when the document contains a hidden
text layer. It can be used in conjunction with
.BR "-sidebar=find" 
to display the text searching interface.

String 
.I text
can be terminated by slash
.BR "" "(" "/" ")"
followed by letters specifying search options.
The following letters are recognized
.br
.TS
center,box;
lfB l
lfB l
lfB l
lfB l
lfB l
lfB l
.
c	Case-sensitive search.
C	Case-insensitive search (default).
w	Search hits start on word boundaries (default).
W	Ignore word boundaries.
r	Regular expression search.
R	String search (default).
.TE
.PP

.TP
.BI "-rotate=(0|90|180|270)"
Rotate the djvu image by the specified angle expressed
in degrees counter-clockwise.

.TP
.BI "-print=(yes|no)"
Enable or disable printing the DjVu document.
Printing is enabled by default.

.TP
.BI "-save=(yes|no)"
Enable or disable saving the DjVu document.
Saving is enabled by default.

.TP
.BI "-passive"
Cause the DjVu image to be displayed in a manner similar 

NEWS

NEW SINCE VERSION 4.10
-------------------
- Support for pinch gestures
- New shortcuts on mac
- Support for high dpi displays
- Spanish translations
- Bug fixes
- Workaround for qt5.5 embed bug

NEW IN VERSION 4.10
-------------------
- Bug fixes: layout, usability, preferences, locale selection
- Works with both Qt4 and Qt5
- Build uses automake/qmake hybrid (instead of autoconf/qmake).
- Layout annotations are honored in multipage mode
- Experimental OpenGL rendering 
- Slideshow mode


NEW IN VERSION 4.9
------------------
- Added language selection in preference.
- Added option to override printing and saving restrictions.
- Automatic reloading of djvu documents when the file is changed.
- New chinese translations.
- Improved network code (needs Qt>=4.4.)
- Remember the last visited position for the last 50 documents.


NEW IN VERSION 4.8
--------------------
- Added option to invert luminance.
- Added shortcut ctrl+g to go to the page combo.
- Fixed export issues under win32.
- Mouse wheel can be configured to control the zoom.
- Animated page movements.


NEW IN VERSION 4.7
--------------------
- Fix for crash on quick PgUp PgDn sequences.
- Fix scrollbars in continuous mode.
- Fix plugin for some webkit based browsers.


NEW IN VERSION 4.6
--------------------
- Show total page number in page combo and page info box.
- Decode percent encoding in annotation and outline hyperlinks.
- Properly handle non-ascii filenames in export and save dialogs.
- Subtle changes to PgUp/PgDn/Home/End/Space/Backspace view changes.
- Fixed suffix management in export/save/print dialogs.


NEW IN VERSION 4.5
--------------------
- Fixed priority of annotation options.
- Fixed plugin for recent versions of gcc.
- Added showposition in advanced option copyUrl.
- New advanced mode to display hidden text.
- Lens display annotations and hidden text.
- Nsdejavu can use the XEmbed protocol when available.
- Plugin is scriptable from JavaScript.


NEW IN VERSION 4.4
--------------------
- Full session management under X11.
- Added advanced feature check box.
- Added advanced command to copy urls, outline, etc.
- New translation (russian)
- History in the find dialog.
- Option to display hidden text in status bar.
- Option to display the first page alone in side-by-side mode.
- Option to display side-by-side in right-to-left mode.
- Drag'n drop files into the viewer.
- Support for Lizardtech's plugin option -showposition.


NEW IN VERSION 4.3
------------------
- Undo/Redo
- New translations (german,czech)
- Fixed double free bug in plugin mode.


NEW IN VERSION 4.2
------------------
- Better support of Qt/Mac
- Escape can be used to exit full screen mode
- Recent files menu
- Open Location


NEW IN VERSION 4.1
------------------
- Fixes for Qt 4.2 and 4.3.
- Fixes to enable compilation on Windows.
- TIFF and PDF export.
- New translation (ukrainian)


NEW IN VERSION 4.0
------------------
- Initial release




src/djview.cpp
//C-  -*- C++ -*-
//C- -------------------------------------------------------------------
//C- DjView4
//C- Copyright (c) 2006-  Leon Bottou
//C-
//C- This software is subject to, and may be distributed under, the
//C- GNU General Public License, either version 2 of the license,
//C- or (at your option) any later version. The license should have
//C- accompanied the software or you may obtain a copy of the license
//C- from the Free Software Foundation at http://www.fsf.org .
//C-
//C- This program is distributed in the hope that it will be useful,
//C- but WITHOUT ANY WARRANTY; without even the implied warranty of
//C- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//C- GNU General Public License for more details.
//C-  ------------------------------------------------------------------

#if AUTOCONF
# include "config.h"
#endif

#include <libdjvu/ddjvuapi.h>
#include <libdjvu/miniexp.h>
#include <locale.h>
#include <stdio.h>

#include "qdjvu.h"
#include "qdjview.h"
#include "qdjvuwidget.h"
#include "djview.h"
#ifndef NPDJVU
# include "qdjviewplugin.h"
#endif

#include <QtGlobal>
#include <QApplication>
#include <QByteArray>
#include <QDebug>
#include <QDir>
#include <QFileOpenEvent>
#include <QLibraryInfo>
#include <QLocale>
#include <QRegExp>
#include <QSettings>
#include <QSessionManager>
#include <QString>
#include <QStringList>
#include <QTranslator>

#if defined(Q_OS_WIN32)
# include <mbctype.h>
#endif


#ifdef QT_NO_DEBUG
static bool verbose = false;
#else
static bool verbose = true;
#endif
static bool appReady = false;

static void 
qt4MessageHandler(QtMsgType type, const char *msg)
{
  switch (type) 
    {
    case QtFatalMsg:
      fprintf(stderr,"djview fatal error: %s\n", msg);
      abort();
    case QtCriticalMsg:
      fprintf(stderr,"djview critical error: %s\n", msg);
      break;
    case QtWarningMsg:
      if (verbose || !appReady)
        fprintf(stderr,"djview: %s\n", msg);
      break;
    default:
      if (verbose || !appReady)
        fprintf(stderr,"%s\n", msg);
      break;
    }
}

#if QT_VERSION >= 0x50000
static void 
qt5MessageHandler(QtMsgType type, const QMessageLogContext&, const QString &msg)
{
  QByteArray bmsg = msg.toLocal8Bit();
  qt4MessageHandler(type, bmsg.data());
}
#endif


static void
message(QString string, bool prefix=true)
{
  QByteArray m = string.toLocal8Bit();
  if (prefix)
    fprintf(stderr, "djview: ");
  fprintf(stderr, "%s\n", (const char*) m);
}

static void
message(QStringList sl)
{
  foreach (QString s, sl)
    message(s);
}

QDjViewApplication::QDjViewApplication(int &argc, char **argv)
  : QApplication(argc, argv), 
    context(argv[0])
{
  // Message handler
#if QT_VERSION >= 0x50000
  qInstallMessageHandler(qt5MessageHandler);
#else
  qInstallMsgHandler(qt4MessageHandler);
#endif
  
  // Locale should not mess with printf
  // We do this again because old libdjvulibre
  // did not correctly set LC_NUMERIC.
#ifdef LC_ALL
  ::setlocale(LC_ALL, "");
# ifdef LC_NUMERIC
  ::setlocale(LC_NUMERIC, "C");
# endif
#endif

  // Parse arguments using OEM code page under windows.
#if defined(Q_OS_WIN32)
  _setmbcp(_MB_CP_OEM);
#endif
  // Mac/Cocoa bug workaround
#if defined(Q_OS_DARWIN) && defined(QT_MAC_USE_COCOA) && QT_VERSION<0x40503
  extern void qt_mac_set_native_menubar(bool);
  qt_mac_set_native_menubar(false);
#endif
  
  // Enable highdpi pixmaps
#if QT_VERSION >= 0x50200
  setAttribute(Qt::AA_UseHighDpiPixmaps, true);
#endif
  
  // Wire session management signals
  connect(this, SIGNAL(saveStateRequest(QSessionManager&)),
          this, SLOT(saveSessionState(QSessionManager&)) );
  
  // Install translators
  QStringList langs = getTranslationLangs();
  QTranslator *djviewTrans = new QTranslator(this);
  QTranslator *qtTrans = new QTranslator(this);
  if (loadTranslators(langs, djviewTrans, qtTrans))
    {
      installTranslator(qtTrans);
      installTranslator(djviewTrans);
    }
  else
    {
      delete qtTrans;
      delete djviewTrans;
    }
  // Application is ready
  appReady = true;
}


QDjView * 
QDjViewApplication::newWindow()
{
  if (lastWindow && !lastWindow->getDocument())
    return lastWindow;
  QDjView *main = new QDjView(context);
  main->setAttribute(Qt::WA_DeleteOnClose);
  lastWindow = main;
  return main;
}


static void
addDirectory(QStringList &dirs, QString path)
{
  QString dirname = QDir::cleanPath(path);
  if (! dirs.contains(dirname))
    dirs << dirname;
}


QStringList 
QDjViewApplication::getTranslationDirs()
{
  if (translationDirs.isEmpty())
    {
      QStringList dirs;
      QDir dir = applicationDirPath();
      QString dirPath = dir.canonicalPath();
      addDirectory(dirs, dirPath);
#ifdef DIR_DATADIR
      QString datadir = DIR_DATADIR;
      addDirectory(dirs, datadir + "/djvu/djview4");
      addDirectory(dirs, datadir + "/djview4");
#endif
#ifdef Q_OS_DARWIN
      addDirectory(dirs, dirPath + "/Resources/$LANG.lproj");
      addDirectory(dirs, dirPath + "/../Resources/$LANG.lproj");
      addDirectory(dirs, dirPath + "/../../Resources/$LANG.lproj");
#endif
      addDirectory(dirs, dirPath + "/share/djvu/djview4");
      addDirectory(dirs, dirPath + "/share/djview4");
      addDirectory(dirs, dirPath + "/../share/djvu/djview4");
      addDirectory(dirs, dirPath + "/../share/djview4");
      addDirectory(dirs, dirPath + "/../../share/djvu/djview4");
      addDirectory(dirs, dirPath + "/../../share/djview4");
      addDirectory(dirs, "/usr/share/djvu/djview4");
      addDirectory(dirs, "/usr/share/djview4");
      addDirectory(dirs, QLibraryInfo::location(QLibraryInfo::TranslationsPath));
      translationDirs = dirs;
    }
  return translationDirs;
}


static void
addLang(QStringList &langs, QString s)
{
  if (s.size() > 0)
    {
      s = s.replace(QChar('-'), QChar('_'));
      if (! langs.contains(s))
        langs << s;
#ifdef Q_OS_DARWIN
      if (s == "zh_Hans" && ! langs.contains("zh_CN"))
        langs << "zh_CN";
      if (s == "zh_Hant" && ! langs.contains("zh_TW"))
        langs << "zh_TW";
#endif
    }
}


QStringList 
QDjViewApplication::getTranslationLangs()
{
  if (translationLangs.isEmpty())
    {
      QStringList langs; 
      addLang(langs, QSettings().value("language").toString());
      QString varLanguage = QString::fromLocal8Bit(::getenv("LANGUAGE"));
      foreach(QString lang, varLanguage.split(":", QString::SkipEmptyParts))
        addLang(langs, lang);
#ifdef LC_MESSAGES
      addLang(langs, QString::fromLocal8Bit(::setlocale(LC_MESSAGES, 0)));
#endif
      addLang(langs, QString::fromLocal8Bit(::getenv("LANG")));
#ifdef Q_OS_DARWIN
      QSettings g(".", "globalPreferences");
      foreach (QString lang, g.value("AppleLanguages").toStringList())
        addLang(langs, lang);
#endif
      addLang(langs, QLocale::system().name());
      translationLangs = langs;
    }
  return translationLangs;
}


static bool loadOneTranslator(QTranslator *trans, 
                              QString name, QString lang, QStringList dirs)
{
  QString llang = lang.toLower();
  foreach (QString dir, dirs)
    {
      dir = dir.replace(QRegExp("\\$LANG(?!\\w)"), lang);
      QDir qdir(dir);
      if (qdir.exists())
        {
          if (trans->load(name + "_" + lang, dir, "_.-"))
            return true;
          if (lang != llang && trans->load(name + "_" + llang, dir, "_.-"))
            return true;
        }
    }
  return false;
}

bool
QDjViewApplication::loadTranslators(QStringList langs, 
                                    QTranslator *dTrans, QTranslator *qTrans)
{
  QStringList dirs = getTranslationDirs();
  foreach (QString lang, langs)
    {
      bool okay = true;
      if (okay && dTrans && !loadOneTranslator(dTrans, "djview", lang, dirs))
        okay = false;
      if (okay && qTrans && !loadOneTranslator(qTrans, "qt", lang, dirs))
        okay = false;
      if (okay || lang.startsWith("en_") || lang == "en")
        return okay;
    }
  return false;
}


bool 
QDjViewApplication::event(QEvent *ev)
{
  if (ev->type() == QEvent::FileOpen)
    {
      QString name = static_cast<QFileOpenEvent*>(ev)->file();
      QDjView *main = newWindow();
      if (main->open(name))
        {
          main->show();
        }
      else
        {
          message(tr("cannot open '%1'.").arg(name));
          delete main;
        }
      return true;
    }
  else if (ev->type() == QEvent::Close)
    {
      closeAllWindows();
    }
  return QApplication::event(ev);
}

void 
QDjViewApplication::saveSessionState(QSessionManager &sm)
{
  int n = 0;
  QSettings s;
  foreach(QWidget *w, topLevelWidgets())
    {
      QDjView *djview = qobject_cast<QDjView*>(w);
      if (djview && !djview->objectName().isEmpty() &&
          djview->getViewerMode() == QDjView::STANDALONE )
        {
          if (++n == 1)
            {
              QStringList discard = QStringList(applicationFilePath());
              discard << QLatin1String("-discard") << sessionId();
              sm.setDiscardCommand(discard);
              QStringList restart = QStringList(applicationFilePath());
              restart << QLatin1String("-session");
              restart << sessionId() + "_" + sessionKey();
              sm.setRestartCommand(restart);
              QString id = QLatin1String("session_") + sessionId();
              s.sync();
              s.remove(id);
              s.beginGroup(id);
            }
          s.beginGroup(QString::number(n));
          djview->saveSession(&s);
          s.endGroup();
        }
    }
  if (n > 0)
    {
      s.setValue("windows", n);
      s.endGroup();
      s.sync();
    }
}


#ifndef NPDJVU

static void 
usage()
{
  QString msg = QDjViewApplication::tr
    ("Usage: djview [options] [filename-or-url]\n"
     "Common options include:\n"
     "-help~~~Prints this message.\n"
     "-verbose~~~Prints all warning messages.\n"
     "-display <xdpy>~~~Select the X11 display <xdpy>.\n"
     "-geometry <xgeom>~~~Select the initial window geometry.\n"
     "-font <xlfd>~~~Select the X11 name of the main font.\n"
     "-style <qtstyle>~~~Select the QT user interface style.\n"
     "-fullscreen, -fs~~~Start djview in full screen mode.\n"
     "-page=<page>~~~Jump to page <page>.\n"
     "-zoom=<zoom>~~~Set zoom factor.\n"
     "-continuous=<yn>~~~Set continuous layout.\n"
     "-sidebyside=<yn>~~~Set side-by-side layout.\n" );
  
  // align tabs
  QStringList opts = msg.split("\n");
  int tab = 0;
  foreach (QString s, opts)
    tab = qMax(tab, s.indexOf("~~~"));
  foreach (QString s, opts)
    {
      int pos = s.indexOf("~~~");
      if (pos >= 0)
        s = QString(" %1  %2").arg(s.left(pos), -tab).arg(s.mid(pos+3));
      message(s, false);
    }
  exit(10);
}


int 
main(int argc, char *argv[])
{
  // Application data
  QApplication::setOrganizationName(DJVIEW_ORG);
  QApplication::setOrganizationDomain(DJVIEW_DOMAIN);
  QApplication::setApplicationName(DJVIEW_APP);
  
  // Message handler
#if QT_VERSION >= 0x50000
  qInstallMessageHandler(qt5MessageHandler);
#else
  qInstallMsgHandler(qt4MessageHandler);
#endif

  // Set verbose flag as early as possible
#ifdef Q_OS_UNIX
  const char *s = ::getenv("DJVIEW_VERBOSE");
  if (s && strcmp(s,"0"))
    verbose = true;
#endif
  for (int i=1; i<argc; i++)
    {
      char *s = argv[i];
      while (s && *s=='-') 
        s++;
      if (! strcmp(s, "verbose"))
        verbose = true;
    }
  
  // Color specification 
  // (cause XRender errors under many versions of Qt4/X11)
#ifndef Q_WS_X11 
  QApplication::setColorSpec(QApplication::ManyColor);
#endif
  
  // Plugin mode
#if WITH_DJVIEWPLUGIN
  if (argc==2 && !strcmp(argv[1],"-netscape"))
    {
      verbose = true;
      QDjViewPlugin dispatcher(argv[0]);
      return dispatcher.exec();
    }
#endif
  
  // Discard session
  if (argc==3 && !strcmp(argv[1],"-discard"))
    {
      QSettings s;
      s.remove(QLatin1String("session_")+QLatin1String(argv[2]));
      s.sync();
      return 0;
    }
  
  // Create application
  QDjViewApplication app(argc, argv);

  // Restore session
  if (app.isSessionRestored())
    {
      QSettings s;
      s.beginGroup(QLatin1String("session_") + app.sessionId());
      int windows = s.value("windows").toInt();
      if (windows > 0)
        {
          for (int n=1; n<=windows; n++)
            {
              QDjView *w = new QDjView(*app.djvuContext());
              w->setAttribute(Qt::WA_DeleteOnClose);
              s.beginGroup(QString::number(n));
              w->restoreSession(&s);
              s.endGroup();
              w->show();
            }
          return app.exec();
        }
    }
  
  // Process command line
  QStringList args;
  QDjView *main = app.newWindow();
  while (argc > 1 && argv[1][0] == '-')
    {
      QString arg = QString::fromLocal8Bit(argv[1]).replace(QRegExp("^-+"),"");
      QString key = arg.section(QChar('='),0,1);
      if (arg == "help")
        usage();
      else if (arg == "verbose")
        verbose = true;
      else if (arg == "quiet")
        verbose = false;
      else if (arg == "fix")
        message(QApplication::tr("Option '-fix' is deprecated."));
      else 
        args += arg;
      argc --;
      argv ++;
    }
  if (argc > 2)
    usage();

  // Open file
  if (argc > 1)
    {
      QString name = QString::fromLocal8Bit(argv[1]);
      bool okay = true;
      if (name.contains(QRegExp("^[a-zA-Z]{3,8}:/")))
        okay = main->open(QUrl(name));
      else
        okay = main->open(name);
      if (! okay)
        {
          message(QDjView::tr("cannot open '%1'.").arg(argv[1]));
          exit(10);
        }
    }
  
  // Process options and go
  foreach(QString arg, args)
    message(main->parseArgument(arg));
  main->show();
  return app.exec();
}

#endif


/* -------------------------------------------------------------
   Local Variables:
   c++-font-lock-extra-types: ( "\\sw+_t" "[A-Z]\\sw*[a-z]\\sw*" )
   End:
   ------------------------------------------------------------- */
